consumer/kafka/src/main/java/com/spotify/heroic/consumer/kafka/OffsetCommitThread.java

```diff
@@ -0,0 +1,182 @@
+    protected final ResolvableFuture<Void> hasStopped;
+
+    private final AtomicBoolean currentlyRunning = new AtomicBoolean(false);
+    private Stopwatch fullOperationWatch;
+
+    public OffsetCommitThread(
+        final AsyncFramework async, final List<ConsumerThread> consumerThreads,
+        final ConsumerConnector consumerConnector
+    ) {
+        super(OffsetCommitThread.class.getCanonicalName());
+
+        this.name = OffsetCommitThread.class.getCanonicalName();
+        this.async = async;
+        this.consumerThreads = consumerThreads;
+        this.consumerConnector = consumerConnector;
+        this.hasStopped = async.future();
+        this.fullOperationWatch = null;
+    }
+
+    @Override
+    public void run() {
+        log.info("{}: Starting thread", name);
+
+        try {
+            guardedRun();
+        } catch (final Throwable e) {
+            log.error("{}: Error in thread", name, e);
+            hasStopped.fail(e);
+            return;
+        }
+
+        log.info("{}: Stopping thread", name);
+        hasStopped.resolve(null);
+        return;
+    }
+
+    public AsyncFuture<Void> shutdown() {
+        shouldStop.countDown();
+        return hasStopped;
+    }
+
+    private void guardedRun() throws Exception {
+        shouldStop.await(INITIAL_SLEEP, TimeUnit.SECONDS);
+        while (shouldStop.getCount() != 0) {
+            runOnce();
+
+            shouldStop.await(COMMIT_INTERVAL, TimeUnit.SECONDS);
+        }
+    }
+
+    private void runOnce() {
+        log.info("###### Preparing kafka-offset-commit");
+        fullOperationWatch = Stopwatch.createStarted();
+        boolean isAlreadyRunning = currentlyRunning.getAndSet(true);
+        if (isAlreadyRunning) {
+            // Already running asynchronously - will skip this round
+            log.info("A Kafka offset commit operation is already underway");
+            return;
+        }
+
+        List<AsyncFuture<Void>> futures = new ArrayList<>();
+        for (ConsumerThread t : consumerThreads) {
+            futures.add(t.pauseConsumption());
+        }
+        AsyncFuture<Void> allThreadsIsPaused = async.collectAndDiscard(futures);
+        allThreadsIsPaused.onResolved((voidValue) -> waitForOutstandingWrites());
+        allThreadsIsPaused.onFailed(throwable -> failed());
+        allThreadsIsPaused.onCancelled(() -> cancelled());
+    }
+
+    private void waitForOutstandingWrites() {
+        // At this point all ConsumerThread's are paused
+        log.info("All ConsumerThreads are paused (took " +
+            fullOperationWatch.elapsed(TimeUnit.MILLISECONDS) + " ms)");
+        Stopwatch waitForWritesWatch = Stopwatch.createStarted();
+
+        // Now, find all outstanding writes and wait for them to complete
+        List<AsyncFuture<Void>> outstandingWrites = new ArrayList<>();
+        for (ConsumerThread t : consumerThreads) {
+            AsyncFuture<Void> maybe = t.getLatestConsumptionFuture();
```

Comment:
It is important to note that updates to `latestConsumptionFuture` might not be visible to other threads.
In order to guarantee that writes to the `latestIngestionFuture` field are visible across threads we must establish a happens-before relationship. This can be done through synchronization, or with the use of `volatile` or something equivalent - like writing/reading an AtomicInteger.

For details, see: https://www.cs.umd.edu/~pugh/java/memoryModel/jsr-133-faq.html